Shared/Realm.Shared/RealmCollectionBase.cs

```diff
@@ -0,0 +1,304 @@
+        }
+
+        private void UnsubscribeFromNotifications(NotificationCallbackDelegate callback)
+        {
+            if (_callbacks.Remove(callback) &&
+                _callbacks.Count == 0)
+            {
+                UnsubscribeFromNotifications();
+            }
+        }
+
+        private void SubscribeForNotifications()
+        {
+            Debug.Assert(_notificationToken == null, "_notificationToken must be null before subscribing.");
+
+            var managedResultsHandle = GCHandle.Alloc(this);
+            var token = new NotificationTokenHandle(Handle);
+            var tokenHandle = Handle.AddNotificationCallback(GCHandle.ToIntPtr(managedResultsHandle), RealmCollectionNativeHelper.NotificationCallback);
+
+            RuntimeHelpers.PrepareConstrainedRegions();
+            try
+            {
+            }
+            finally
+            {
+                token.SetHandle(tokenHandle);
+            }
+
+            _notificationToken = token;
+        }
+
+        private void UnsubscribeFromNotifications()
+        {
+            Debug.Assert(_notificationToken != null, "_notificationToken must not be null to unsubscribe.");
+
+            _notificationToken?.Dispose();
+            _notificationToken = null;
+        }
+
+        #region INotifyCollectionChanged
+
+        private void OnChange(RealmCollectionBase<T> sender, ChangeSet change, Exception error)
+        {
+            if (error != null)
+            {
+                Realm.NotifyError(error);
+            }
+            else if (change != null)
+            {
+                IList removedItems;
+                int removedStartIndex;
+                var raiseRemoved = TryGetConsecutive(change.DeletedIndices, _ => default(T), out removedItems, out removedStartIndex);
+
+                IList addedItems;
+                int addedStartIndex;
+                var raiseAdded = TryGetConsecutive(change.InsertedIndices, ItemAtIndex, out addedItems, out addedStartIndex);
+
+                if (raiseAdded || raiseRemoved)
+                {
+                    if ((raiseAdded && raiseRemoved) ||
+                        (raiseAdded && addedItems == null) ||
+                        (raiseRemoved && removedItems == null))
+                    {
+                        RaiseCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+                        return;
+                    }
+
+                    if (removedItems != null)
+                    {
+                        RaiseCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, removedItems, removedStartIndex));
+                    }
+
+                    if (addedItems != null)
+                    {
+                        RaiseCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, addedItems, addedStartIndex));
+                    }
+                }
```

Comment:
The initial notification, which will be fired as soon as the query is run on the worker thread.

Because the previous implementation used a cache that eagerly evaluated the query we didn't care about the initial notification because the intermediate collection had already pulled in everything in the `Results` object at the time of its creation.

In practice, anyone who cares enough to subscribe to `CollectionChanged` is likely to have attempted to enumerate the `Results` object anyway which will eagerly run its query, so we might not care about the initial notification.